src/autocomplete/zsh.ts

```diff
@@ -0,0 +1,400 @@
+
+      // skip hidden flags
+      if (f.hidden) continue
+
+      f.summary = sanitizeSummary(f.summary || f.description)
+
+      let flagSpec = ''
+
+      if (f.type === 'option') {
+        if (f.char) {
+          if (f.multiple) {
+            // this flag can be present multiple times on the line
+            flagSpec += `"*"{-${f.char},--${f.name}}`
+          } else {
+            flagSpec += `"(-${f.char} --${f.name})"{-${f.char},--${f.name}}`
+          }
+
+          flagSpec += `"[${f.summary}]`
+
+          if (f.options) {
+            flagSpec += `:${f.name} options:(${f.options?.join(' ')})"`
+          } else {
+            flagSpec += ':file:_files"'
+          }
+        } else {
+          if (f.multiple) {
+            // this flag can be present multiple times on the line
+            flagSpec += '"*"'
+          }
+
+          flagSpec += `--${f.name}"[${f.summary}]:`
+
+          if (f.options) {
+            flagSpec += `${f.name} options:(${f.options.join(' ')})"`
+          } else {
+            flagSpec += 'file:_files"'
+          }
+        }
+      } else if (f.char) {
+        // Flag.Boolean
+        flagSpec += `"(-${f.char} --${f.name})"{-${f.char},--${f.name}}"[${f.summary}]"`
+      } else {
+        // Flag.Boolean
+        flagSpec += `--${f.name}"[${f.summary}]"`
+      }
+
+      flagSpec += ' \\\n'
+      argumentsBlock += flagSpec
+    }
+    // complete files if `-` is not present on the current line
+    argumentsBlock += '"*: :_files"'
+
+    return argumentsBlock
+  }
+
+  private genZshValuesBlock(subArgs: {id: string; summary?: string}[]): string {
+    let valuesBlock = '_values "completions" \\\n'
+
+    subArgs.forEach(subArg => {
+      valuesBlock += `"${subArg.id}[${subArg.summary}]" \\\n`
+    })
+
+    return valuesBlock
+  }
+
+  private genZshTopicCompFun(id: string): string {
+    const coTopics: string[] = []
+
+    for (const topic of this.topics) {
+      for (const cmd of this.commands) {
+        if (topic.name === cmd.id) {
+          coTopics.push(topic.name)
+        }
+      }
+    }
+
+    const flagArgsTemplate = '        "%s")\n          %s\n        ;;\n'
+
+    const underscoreSepId = id.replace(/:/g, '_')
+    const depth = id.split(':').length
+
+    const isCotopic = coTopics.includes(id)
+
+    if (isCotopic) {
+      const compFuncName = `${this.config.bin}_${underscoreSepId}`
+
+      const coTopicCompFunc =
+`_${compFuncName}() {
+  _${compFuncName}_flags() {
+    local context state state_descr line
+    typeset -A opt_args
+
+    ${this.genZshFlagArgumentsBlock(this.commands.find(c => c.id === id)?.flags)}
+  }
+
+  local context state state_descr line
+  typeset -A opt_args
+
+  _arguments -C "1: :->cmds" "*::arg:->args"
+
+  case "$state" in
+    cmds)
+      if [[ "\${words[CURRENT]}" == -* ]]; then
+        _${compFuncName}_flags
+      else
+%s
+      fi
+      ;;
+    args)
+      case $line[1] in
+%s
+      *)
+        _${compFuncName}_flags
```

Comment:
`sf data query --json -<tab>`

`line[1]` is the last word on the current line, if it matches any valid command of the cotopic it will suggest flags valid to that command.

`*)` is the "default" block of the switch statement, if `line[1]` doesn't match any valid command of the cotopic then keep suggesting flags of the cotopic.